internal/pkg/oras/oras.go

```diff
@@ -0,0 +1,183 @@
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	"github.com/sylabs/singularity/pkg/image"
+)
+
+const (
+	// SifDefaultTag is the tag to use when a tag is not specified
+	SifDefaultTag = "latest"
+
+	// SifConfigMediaType is the config descriptor mediaType
+	SifConfigMediaType = "application/vnd.sylabs.sif.config.v1+json"
+
+	// SifLayerMediaType is the mediaType for the "layer" which contains the actual SIF file
+	SifLayerMediaType = "appliciation/vnd.sylabs.sif.layer.tar"
+)
+
+// DownloadImage downloads a SIF image specified by an oci reference to a file using the included credentials
+func DownloadImage(imagePath, ref string, ociAuth *ocitypes.DockerAuthConfig) error {
+	ref = strings.TrimPrefix(ref, "//")
+
+	spec, err := reference.Parse(ref)
+	if err != nil {
+		return fmt.Errorf("unable to parse oci reference: %s", err)
+	}
+
+	// append default tag if no object exists
+	if spec.Object == "" {
+		spec.Object = SifDefaultTag
+		sylog.Infof("No tag or digest found, using default: %s", SifDefaultTag)
+	}
+
+	credFn := func(_ string) (string, string, error) {
+		if ociAuth != nil {
+			return ociAuth.Username, ociAuth.Password, nil
+		}
+
+		return "", "", nil
+	}
+	resolver := docker.NewResolver(docker.ResolverOptions{Credentials: credFn})
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get working directory: %s", err)
+	}
+
+	store := content.NewFileStore(wd)
+	defer store.Close()
+
+	store.AllowPathTraversalOnWrite = true
+	store.DisableOverwrite = true
+
+	allowedMediaTypes := oras.WithAllowedMediaTypes([]string{SifLayerMediaType})
+	handlerFunc := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if desc.MediaType == SifLayerMediaType {
+			// Ensure descriptor is of a single file
+			// AnnotationUnpack indicates that the descriptor is of a directory
+			if desc.Annotations[content.AnnotationUnpack] == "true" {
+				return nil, fmt.Errorf("descriptor is of a bundled directory, not a SIF image")
+			}
+			nameOld, _ := content.ResolveName(desc)
+			_ = store.MapPath(nameOld, imagePath)
+		}
+		return nil, nil
+	}
+	pullHandler := oras.WithPullBaseHandler(images.HandlerFunc(handlerFunc))
+
+	_, _, err = oras.Pull(orasctx.Background(), resolver, spec.String(), store, allowedMediaTypes, pullHandler)
+	if err != nil {
+		return fmt.Errorf("unable to pull from registry: %s", err)
+	}
+
+	// ensure that we have downloaded a SIF
+	if err := EnsureSIF(imagePath); err != nil {
+		// remove whatever we downloaded if it is not a SIF
+		os.RemoveAll(imagePath)
+		return err
+	}
+
+	// ensure container is executable
+	if err := os.Chmod(imagePath, 0755); err != nil {
+		return fmt.Errorf("unable to set image perms: %s", err)
+	}
+
+	return nil
+}
+
+// EnsureSIF checks for a SIF image at filepath and returns an error if it is not, or an error is encountered
+func EnsureSIF(filepath string) error {
```

Comment:
Any reason why this is in `oras.go`? It seems to me that it is independent from ORAS.

Reply:
No reason specifically. Not sure if this is something we want in `pkg/image/image.go`. Maybe this could go into `internal/pkg/util`?

Reply:
After thinking a bit more on this, I'm `OrasPush()` into this `oras` package and make `EnsureSIF()` an internal utility function